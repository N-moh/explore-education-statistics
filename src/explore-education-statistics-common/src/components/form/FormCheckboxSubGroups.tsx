import { Overwrite } from '@common/types/util';
import camelCase from 'lodash/camelCase';
import React, { ChangeEvent } from 'react';
import FormCheckboxGroup, {
  CheckboxOption,
  FormCheckboxGroupProps,
} from './FormCheckboxGroup';
import FormFieldset, { FormFieldsetProps } from './FormFieldset';

export type FormCheckboxSubGroupsProps = Overwrite<
  FormCheckboxGroupProps,
  {
    options: {
      id?: string;
      legend: string;
      options: CheckboxOption[];
    }[];
    onAllChange?: (
      event: ChangeEvent<HTMLInputElement>,
      options: CheckboxOption[],
    ) => void;
  }
>;

const FormCheckboxSubGroups = ({
  legendHidden,
  legendSize = 's',
  legend,
  hint,
  error,
  ...props
}: FormCheckboxSubGroupsProps) => {
  const { id, onAllChange, options } = props;
  const fieldsetProps: FormFieldsetProps = {
    id,
    legend,
    legendSize,
    legendHidden,
    hint,
    error,
  };

  return (
    <FormFieldset {...fieldsetProps}>
      {options.map(optionGroup => {
        return (
          <FormCheckboxGroup
            {...props}
            id={
              optionGroup.id
                ? optionGroup.id
                : `${id}-${camelCase(optionGroup.legend)}`
            }
            key={optionGroup.legend}
            legend={optionGroup.legend}
            legendSize="s"
            options={optionGroup.options}
            onAllChange={event => {
              if (onAllChange) {
                onAllChange(event, optionGroup.options);
              }
            }}
          />
        );
      })}
    </FormFieldset>
  );
};

export default FormCheckboxSubGroups;
