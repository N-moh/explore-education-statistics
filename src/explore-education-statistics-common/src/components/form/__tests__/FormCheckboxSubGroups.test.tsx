import React from 'react';
import { render } from 'react-testing-library';
import FormCheckboxSubGroups from '../FormCheckboxSubGroups';

describe('FormCheckboxSubGroups', () => {
  test('renders groups of checkboxes in correct order', () => {
    const { container, getAllByLabelText, getAllByText } = render(
      <FormCheckboxSubGroups
        id="test-checkboxes"
        name="test"
        value={[]}
        legend="Test checkboxes"
        options={[
          {
            legend: 'Group A',
            options: [
              { label: 'Checkbox 1', value: '1' },
              { label: 'Checkbox 2', value: '2' },
            ],
          },
          {
            legend: 'Group B',
            options: [
              { label: 'Checkbox 3', value: '3' },
              { label: 'Checkbox 4', value: '4' },
            ],
          },
        ]}
      />,
    );

    const groups = getAllByText(/Group/);

    expect(groups[0]).toHaveTextContent('Group A');
    expect(groups[1]).toHaveTextContent('Group B');

    const checkboxes = getAllByLabelText(/Checkbox/);

    expect(checkboxes[0]).toHaveAttribute('value', '1');
    expect(checkboxes[1]).toHaveAttribute('value', '2');
    expect(checkboxes[2]).toHaveAttribute('value', '3');
    expect(checkboxes[3]).toHaveAttribute('value', '4');

    expect(container.innerHTML).toMatchSnapshot();
  });

  test('renders `Unselect all options` buttons when `selectAll` is true', () => {
    const { container, getAllByText } = render(
      <FormCheckboxSubGroups
        value={[]}
        id="test-checkboxes"
        name="test-checkboxes"
        legend="Test checkboxes"
        selectAll
        options={[
          {
            legend: 'Group A',
            options: [
              { label: 'Checkbox 1', value: '1' },
              { label: 'Checkbox 2', value: '2' },
            ],
          },
          {
            legend: 'Group B',
            options: [
              { label: 'Checkbox 3', value: '3' },
              { label: 'Checkbox 4', value: '4' },
              { label: 'Checkbox 5', value: '5' },
            ],
          },
        ]}
      />,
    );

    const selectAllCheckboxes = getAllByText(
      /Select all/,
    ) as HTMLButtonElement[];

    expect(selectAllCheckboxes[0]).toHaveTextContent('Select all 2 options');
    expect(selectAllCheckboxes[1]).toHaveTextContent('Select all 3 options');

    expect(container.innerHTML).toMatchSnapshot();
  });

  test('renders `Unselect all options` buttons when all options are pre-checked', () => {
    const { getAllByText } = render(
      <FormCheckboxSubGroups
        value={['1', '2', '3', '4', '5']}
        id="test-checkboxes"
        name="test-checkboxes"
        legend="Test checkboxes"
        selectAll
        options={[
          {
            legend: 'Group A',
            options: [
              { label: 'Checkbox 1', value: '1' },
              { label: 'Checkbox 2', value: '2' },
            ],
          },
          {
            legend: 'Group B',
            options: [
              { label: 'Checkbox 3', value: '3' },
              { label: 'Checkbox 4', value: '4' },
              { label: 'Checkbox 5', value: '5' },
            ],
          },
        ]}
      />,
    );

    const selectAllCheckboxes = getAllByText(
      /Unselect all/,
    ) as HTMLInputElement[];

    expect(selectAllCheckboxes[0]).toHaveTextContent('Unselect all 2 options');
    expect(selectAllCheckboxes[1]).toHaveTextContent('Unselect all 3 options');
  });

  test('does not render `Select all options`  when there is only one option', () => {
    const { queryAllByLabelText } = render(
      <FormCheckboxSubGroups
        id="test-checkboxes"
        name="test-checkboxes"
        legend="Test checkboxes"
        selectAll
        value={[]}
        options={[
          {
            legend: 'Group A',
            options: [{ label: 'Checkbox 1', value: '1' }],
          },
          {
            legend: 'Group B',
            options: [{ label: 'Checkbox 3', value: '3' }],
          },
        ]}
      />,
    );

    expect(queryAllByLabelText(/Select all/)).toHaveLength(0);
  });

  test('generates group IDs from group legends if none provided', () => {
    const { container } = render(
      <FormCheckboxSubGroups
        id="test-checkboxes"
        name="test-checkboxes"
        legend="Test checkboxes"
        value={[]}
        options={[
          {
            legend: 'Group A',
            options: [{ label: 'Checkbox 1', value: '1' }],
          },
          {
            legend: 'Group B',
            id: 'custom-group-id',
            options: [{ label: 'Checkbox 3', value: '3' }],
          },
        ]}
      />,
    );

    expect(container.querySelector('#test-checkboxes-groupA')).not.toBeNull();
    expect(container.querySelector('#test-checkboxes-groupB')).toBeNull();
    expect(container.querySelector('#custom-group-id')).not.toBeNull();
  });
});
