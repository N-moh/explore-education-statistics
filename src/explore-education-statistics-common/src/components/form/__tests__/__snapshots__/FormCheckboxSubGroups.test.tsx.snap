// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FormCheckboxSubGroups renders \`Unselect all options\` buttons when \`selectAll\` is true 1`] = `

<div class="govuk-form-group">
  <fieldset class="govuk-fieldset"
            id="test-checkboxes"
  >
    <legend class="govuk-fieldset__legend govuk-fieldset__legend--s">
      Test checkboxes
    </legend>
    <div class="govuk-form-group">
      <fieldset class="govuk-fieldset"
                id="test-checkboxes-groupA"
      >
        <legend class="govuk-fieldset__legend govuk-fieldset__legend--s">
          Group A
        </legend>
        <div class="govuk-checkboxes">
          <button id="test-checkboxes-groupA-all"
                  class="button selectAll"
                  type="button"
          >
            Select all 2 options
          </button>
          <div class="govuk-checkboxes__item">
            <input class="govuk-checkboxes__input"
                   id="test-checkboxes-groupA-1"
                   name="test-checkboxes"
                   type="checkbox"
                   value="1"
            >
            <label class="govuk-label govuk-checkboxes__label"
                   for="test-checkboxes-groupA-1"
            >
              Checkbox 1
            </label>
          </div>
          <div class="govuk-checkboxes__item">
            <input class="govuk-checkboxes__input"
                   id="test-checkboxes-groupA-2"
                   name="test-checkboxes"
                   type="checkbox"
                   value="2"
            >
            <label class="govuk-label govuk-checkboxes__label"
                   for="test-checkboxes-groupA-2"
            >
              Checkbox 2
            </label>
          </div>
        </div>
      </fieldset>
    </div>
    <div class="govuk-form-group">
      <fieldset class="govuk-fieldset"
                id="test-checkboxes-groupB"
      >
        <legend class="govuk-fieldset__legend govuk-fieldset__legend--s">
          Group B
        </legend>
        <div class="govuk-checkboxes">
          <button id="test-checkboxes-groupB-all"
                  class="button selectAll"
                  type="button"
          >
            Select all 3 options
          </button>
          <div class="govuk-checkboxes__item">
            <input class="govuk-checkboxes__input"
                   id="test-checkboxes-groupB-3"
                   name="test-checkboxes"
                   type="checkbox"
                   value="3"
            >
            <label class="govuk-label govuk-checkboxes__label"
                   for="test-checkboxes-groupB-3"
            >
              Checkbox 3
            </label>
          </div>
          <div class="govuk-checkboxes__item">
            <input class="govuk-checkboxes__input"
                   id="test-checkboxes-groupB-4"
                   name="test-checkboxes"
                   type="checkbox"
                   value="4"
            >
            <label class="govuk-label govuk-checkboxes__label"
                   for="test-checkboxes-groupB-4"
            >
              Checkbox 4
            </label>
          </div>
          <div class="govuk-checkboxes__item">
            <input class="govuk-checkboxes__input"
                   id="test-checkboxes-groupB-5"
                   name="test-checkboxes"
                   type="checkbox"
                   value="5"
            >
            <label class="govuk-label govuk-checkboxes__label"
                   for="test-checkboxes-groupB-5"
            >
              Checkbox 5
            </label>
          </div>
        </div>
      </fieldset>
    </div>
  </fieldset>
</div>

`;

exports[`FormCheckboxSubGroups renders groups of checkboxes in correct order 1`] = `

<div class="govuk-form-group">
  <fieldset class="govuk-fieldset"
            id="test-checkboxes"
  >
    <legend class="govuk-fieldset__legend govuk-fieldset__legend--s">
      Test checkboxes
    </legend>
    <div class="govuk-form-group">
      <fieldset class="govuk-fieldset"
                id="test-checkboxes-groupA"
      >
        <legend class="govuk-fieldset__legend govuk-fieldset__legend--s">
          Group A
        </legend>
        <div class="govuk-checkboxes">
          <div class="govuk-checkboxes__item">
            <input class="govuk-checkboxes__input"
                   id="test-checkboxes-groupA-1"
                   name="test"
                   type="checkbox"
                   value="1"
            >
            <label class="govuk-label govuk-checkboxes__label"
                   for="test-checkboxes-groupA-1"
            >
              Checkbox 1
            </label>
          </div>
          <div class="govuk-checkboxes__item">
            <input class="govuk-checkboxes__input"
                   id="test-checkboxes-groupA-2"
                   name="test"
                   type="checkbox"
                   value="2"
            >
            <label class="govuk-label govuk-checkboxes__label"
                   for="test-checkboxes-groupA-2"
            >
              Checkbox 2
            </label>
          </div>
        </div>
      </fieldset>
    </div>
    <div class="govuk-form-group">
      <fieldset class="govuk-fieldset"
                id="test-checkboxes-groupB"
      >
        <legend class="govuk-fieldset__legend govuk-fieldset__legend--s">
          Group B
        </legend>
        <div class="govuk-checkboxes">
          <div class="govuk-checkboxes__item">
            <input class="govuk-checkboxes__input"
                   id="test-checkboxes-groupB-3"
                   name="test"
                   type="checkbox"
                   value="3"
            >
            <label class="govuk-label govuk-checkboxes__label"
                   for="test-checkboxes-groupB-3"
            >
              Checkbox 3
            </label>
          </div>
          <div class="govuk-checkboxes__item">
            <input class="govuk-checkboxes__input"
                   id="test-checkboxes-groupB-4"
                   name="test"
                   type="checkbox"
                   value="4"
            >
            <label class="govuk-label govuk-checkboxes__label"
                   for="test-checkboxes-groupB-4"
            >
              Checkbox 4
            </label>
          </div>
        </div>
      </fieldset>
    </div>
  </fieldset>
</div>

`;
